MiniApp/Classes/MiniAppDownloader.swift

```diff
@@ -23,30 +23,67 @@ class MiniAppDownloader {
+                         switch downloadResult {
+                         case .success:
+                             self.miniAppStorage.cleanVersions(for: appId, differentFrom: versionId, status: self.miniAppStatus)
+
+                             fallthrough
+                         default:
+                             completionHandler(downloadResult)
+                         }
+                     }
+                 case .failure(let error):
+                    self.handleDownloadError(appId: appId, error: error, completionHandler: completionHandler)
+                 }
+             }
+        } else {
             completionHandler(.success(miniAppStoragePath))
-            return
         }
-        self.manifestDownloader.fetchManifest(apiClient: self.miniAppClient, appId: appId, versionId: versionId) { (result) in
-            switch result {
-            case .success(let responseData):
-                self.downloadMiniApp(urls: responseData.manifest, to: miniAppStoragePath) { downloadResult in
-                    switch downloadResult {
-                    case .success:
-                        self.miniAppStorage.cleanVersions(for: appId, differentFrom: versionId, status: self.miniAppStatus)
+    }
 
-                        fallthrough
-                    default:
-                        completionHandler(downloadResult)
-                    }
-                }
-            case .failure(let error):
-                return completionHandler(.failure(error))
+    func handleDownloadError(appId: String, error: Error, completionHandler: @escaping (Result<URL, Error>) -> Void) {
+        let downloadError = error as NSError
+        if Constants.offlineErrorCodeList.contains(downloadError.code) {
+           guard let miniAppPath = self.getCachedMiniApp(appId: appId) else {
+               return completionHandler(.failure(downloadError))
+           }
+           completionHandler(.success(miniAppPath))
+        }
+        return completionHandler(.failure(error))
+    }
+
+    func isMiniAppAlreadyDownloaded(appId: String, versionId: String) -> Bool {
+        if miniAppStatus.isDownloaded(appId: appId, versionId: versionId) {
+            let versionDirectory = FileManager.getMiniAppVersionDirectory(with: appId, and: versionId)
+            var isDirectory: ObjCBool = true
+            if FileManager.default.fileExists(atPath: versionDirectory.path, isDirectory: &isDirectory) {
+                return true
             }
         }
+        return false
+    }
+
+    /// Check if there is any old version of mini app cached for a given app ID
+    /// - Parameter appId: App ID for which the version directory path to be fetched
+    /// - Returns: URL of the available cached version for a given mini app ID
+    func getCachedMiniApp(appId: String) -> URL? {
+        guard let versionDirectory = FileManager.getMiniAppVersionDirectory(usingAppId: appId) else {
```

Comment:
So `FileManager.getMiniAppVersionDirectory` will inference the version id by checking the folder structure under that App ID? But what if there are two version directories? For example, if there were some error where the download failed halfway through downloading a new version. Then `FileManager.getMiniAppVersionDirectory` could return the wrong path?

It seems dangerous to assume that the folder structure will be correct and that it won't have any invalid folders for failed versions. Could you cache the downloaded version in `MiniAppStatus` instead? I think this is what Android is doing.

Reply:
That's right, I guess I need to make changes in `MiniAppStatus`, in which the way it stores the cache information. Will work on this change, thank you 👍